### PR TITLE
Feature signature validation

### PR DESCRIPTION
**Pull request by dmolineus**
See #13

### CODE REVIEW

**Comment by dmolineus** [comment]
With 8613ee1 we get the first working draft for a signature verification.

I decided to use guzzle as http client so far, encapsulated with psr7/17 so it would be easy to switch.

Todos:

- [ ] Use downloader inside of `KeyDownloader`
- [ ] Unit tests

We should make following decisions:

- [ ] Location of the keychain database
- [ ] Decide if option to always trust keys is necessary
- [ ] Decide how interactive mode to trust keys should be implemented (own command or part of run)
- [ ] Decide if dependency of `phar-io/gnupg` should be kept (it's a wrapper for the pgp binary which is used if the gnupg extension is not used)

**Comment by dmolineus** [comment]
This PR is ready for dicussion now.

* Keys can be trusted by being added to the `phpcq.yaml` configuration, by `--acept-keys` or interactively for each key
* Keychain is located in the `phpcq path` so each project uses it's own keychain

We may still diccuss if we drop dependency of `phar-io/gnupg` again (by reimplement the fallback or drop support for non gnupg based environments).

**Comment by discordier** [comment]
Will go over the implementation in a separate review but want to discuss this here first:
> * Keychain is located in the `phpcq path` so each project uses it's own keychain

IMO it would be better to also have support for a central keychain. This way I would not have to accept keys for every project over and over again (despite being the same keys).
The whole sense of GPG validation is to trust a key once and then have everything signed by it trusted.

We might want a dedicated `PHPCQ_HOME` where we store the keychain (and repository auth data as soon as we have it implemented).

**Comment by dmolineus** [comment]
> IMO it would be better to also have support for a central keychain. This way I would not have to accept keys for every project over and over again (despite being the same keys).
> The whole sense of GPG validation is to trust a key once and then have everything signed by it trusted.

This was also my first thought (and I also implemented it first). Following your argumentation we should use the user keychain that seems wrong to me eigther.

I changed my mind. I think that the tool chain should provide a deterministic result. If a central keychain trusts a key on my local machine but on the CI it isn't, it will lead to an error. Running phpcq with the `trust-keys` option isn't a solution for CI. For could drop signature handling completely then.

This the per project based key chain the project developer has the full control of the supported and trusted keys on a project level.

File: src/Command/ExecCommand.php [hunk 1]
```diff
@@ -65,6 +68,14 @@ protected function configure(): void
             'Optional options and arguments to pass to the tool'
         );
 
+        $this->addOption(
```
Comment (discordier) [inline src/Command/ExecCommand.php:71]: As of 70d01cf18cd19be6cce905c6934ecf51006e5f9d, this should not be needed anymore as it is now unused, the changes in this file should get removed from e458f003855556d434dfc2a64aca074d3970793c.

File: src/FileDownloader.php [hunk 2]
```diff
@@ -7,6 +7,7 @@
 use GuzzleHttp\Client;
 use Phpcq\Exception\InvalidHashException;
 use Phpcq\Exception\RuntimeException;
+use function file_get_contents;
```
Comment (discordier) [inline src/FileDownloader.php:10]: This is either superfluous or `file_put_contents` is missing.

File: src/ToolUpdate/UpdateExecutor.php [hunk 3]
```diff
@@ -37,17 +40,30 @@ final class UpdateExecutor
...
         string $phpcqPath,
-        OutputInterface $output
-    )
-    {
-        $this->platform   = $platform;
-        $this->downloader = $downloader;
-        $this->phpcqPath  = $phpcqPath;
-        $this->output     = $output;
+        OutputInterface $output,
+        callable $untrustedKeyStrategy
```
Comment (discordier) [inline src/ToolUpdate/UpdateExecutor.php:59]: Can't we shift the strategy into the verifier? That way we would not couple this class with yet another one.

IMO fetching and accepting keys on the fly makes more sense in the verifier than in the updater even from a logical point of view. WDYT?

File: src/ToolUpdate/UpdateExecutor.php [hunk 4]
```diff
@@ -125,6 +142,9 @@ private function deleteVersion(ToolInformationInterface $tool): void
         if (!$bootstrap instanceof InstalledBootstrap) {
             throw new RuntimeException('Can only remove installed bootstrap files.');
         }
+        if ($signatureUrl = $tool->getSignatureUrl()) {
```
Comment (discordier) [inline src/ToolUpdate/UpdateExecutor.php:145]: Please do not separate the type check of bootstrap from its actual deletion.
```php
if (!$bootstrap instanceof InstalledBootstrap) {
    throw new RuntimeException('Can only remove installed bootstrap files.');
}
$this->deleteFile($bootstrap->getFilePath());
```
belongs together, so code here should be:
```php
if (!$bootstrap instanceof InstalledBootstrap) {
    throw new RuntimeException('Can only remove installed bootstrap files.');
}
$this->deleteFile($bootstrap->getFilePath());

if ($signatureUrl = $tool->getSignatureUrl()) {
    $this->deleteFile($this->phpcqPath . '/' . $signatureUrl);
}
```

File: src/ToolUpdate/UpdateExecutor.php [hunk 5]
```diff
@@ -154,4 +174,35 @@ private function validateHash(string $pathToPhar, ?ToolHash $hash): void
...
-}
\ No newline at end of file
+
+    private function verifySignature(string $pharPath, ToolInformationInterface $tool): ?string
+    {
+        $signatureUrl = $tool->getSignatureUrl();
+        if (null === $signatureUrl) {
+            // TODO: How should we handle missing signatures:
+            // Show a warning or add a configuration how phpcq should deal with it?
+            return null;
```
Comment (discordier) [inline src/ToolUpdate/UpdateExecutor.php:184]: IMO we should handle missing signatures as fatal unless configured otherwise (eg via whitelist in YAML or the like).

File: src/ToolUpdate/UpdateExecutor.php [hunk 6]
```diff
@@ -154,4 +174,35 @@ private function validateHash(string $pathToPhar, ?ToolHash $hash): void
...
+        }
+
+        $signatureName = sprintf('%1$s~%2$s.asc', $tool->getName(), $tool->getVersion());
+        $signaturePath = $this->phpcqPath . '/' . $signatureName;
+        $this->downloader->downloadFileTo($signatureUrl, $signaturePath);
+        $result = $this->verifier->verify(file_get_contents($pharPath),  file_get_contents($signaturePath));
+
+        if ($result->isUntrustedKey() && ($this->untrustedKeyStrategy)($result->getFingerprint(), $tool)) {
```
Comment (discordier) [inline src/ToolUpdate/UpdateExecutor.php:192]: As mentioned above, `untrustedKeyStrategy` handling should be shifted into `verifier` IMO.

File: src/Command/ExecCommand.php (hunk 1 above)
Comment (dmolineus) [reply src/Command/ExecCommand.php:71]: Hm, it was missing when I tried to run the command but looks like it can be skipped now.

File: src/ToolUpdate/UpdateExecutor.php (hunk 3 above)
Comment (dmolineus) [reply src/ToolUpdate/UpdateExecutor.php:59]: IMO I think that the pgp related stuff could be extracted into an own library. But even if we do it, we can use a wrapper verifier handling the strategy.

**Comment by dmolineus** [comment]
Another issues which has to be addressed:

 - [ ] If verification failed the phar keeps download. We should remove the file or event might have a rollback if the update process fails

File: .travis.yml [hunk 7]
```diff
@@ -37,7 +40,7 @@ install:
         || echo dev-${TRAVIS_BRANCH}) \
         travis_retry composer update $COMPOSER_FLAGS --prefer-dist --no-interaction;
     fi
-  - ./phpcq update
+  - ./phpcq update -t ./.phpcq
```
Comment (discordier) [inline .travis.yml:43]: Why this change? When doing so, it must also get changed below in the run statement as otherwise the tools can't be located anymore.

**Comment by discordier** [comment]
> If verification failed the phar keeps download. We should remove the file or event might have a rollback if the update process fails

Rollback is rather a problem, when dealing with phars, bootstrap and signature, as we would have to roll back all of them in this case.

Sidenote: do we also need a `phpcq.lock` file? Might be a wise idea.

File: src/ToolUpdate/UpdateExecutor.php (hunk 3 above)
Comment (dmolineus) [reply src/ToolUpdate/UpdateExecutor.php]: Done with 1dd7239.

File: src/ToolUpdate/UpdateExecutor.php (hunk 6 above)
Comment (dmolineus) [reply src/ToolUpdate/UpdateExecutor.php]: Done.

File: .travis.yml (hunk 7 above)
Comment (dmolineus) [reply .travis.yml:43]: I wanted to enable vendor caching again by using a custom phpcq directory. Applied the the option to the run command as well.

File: src/ToolUpdate/UpdateExecutor.php (hunk 4 above)
Comment (dmolineus) [reply src/ToolUpdate/UpdateExecutor.php:140]: Done with 42a95b6.

File: src/FileDownloader.php (hunk 2 above)
Comment (dmolineus) [reply src/FileDownloader.php:10]: Both is used but only one imported. Time to get our coding standard ready so that `phpcs` will fix it. @baumannsven ;-)

Comment (dmolineus) [reply src/FileDownloader.php:10]: Changed in 7de3179.

File: src/Command/ExecCommand.php (hunk 1 above)
Comment (dmolineus) [reply src/Command/ExecCommand.php]: Fixed with 91f10e0.

File: src/ToolUpdate/UpdateExecutor.php (hunk 5 above)
Comment (dmolineus) [reply src/ToolUpdate/UpdateExecutor.php]: Implemented with ea713cc6bbcca5f0a672cd40e10e25561946a2fd.

**Comment by dmolineus** [comment]
I extracted the gnupg part into an own library and addressed all issues, so it's time for a new review. :-)
